metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/sync/BlockSynchronizer.scala

```diff
@@ -67,6 +70,55 @@ class BlockSynchronizer[F[_]: Sync: Timer, N, A <: Agreement: Block](
       _    <- persist(quorumCertificate.blockHash, path)
     } yield ()
 
+  /** Download the block in the Quorum Certificate without ancestors.
+    *
+    * Return it without being persisted.
+    *
+    * Unlike `sync`, which is expected to be canceled if consensus times out,
+    * or be satisfied by alternative downloads happening concurrently, this
+    * method returns and error if it cannot download the block after a certain
+    * number of attempts, from any of the sources. This is becuause its primary
+    * use is during state syncing where this is the only operation, and if for
+    * any reason the block would be gone from everyone honest members' storage,
+    * we have to try something else.
+    */
+  def getBlockFromQuorumCertificate(
+      sources: NonEmptyVector[A#PKey],
+      quorumCertificate: QuorumCertificate[A]
+  ): F[Either[DownloadFailedException[A], A#Block]] = {
+    val otherSources = sources.filterNot(_ == publicKey).toList
+
+    def loop(
+        alternatives: List[A#PKey]
+    ): F[Either[DownloadFailedException[A], A#Block]] = {
+      alternatives match {
+        case Nil =>
+          new DownloadFailedException(
+            quorumCertificate.blockHash,
+            sources.toVector
+          ).asLeft[A#Block].pure[F]
+
+        case source :: alternatives =>
```

Comment:
Minor: I try to avoid name shadowing as much as possible. This could be named `remaining` or `nextAlternatives`

Reply:
Interesting, I deliberately tried to use shadowing in these loops because it happened to me that I accidentally reused the input variable which I didn't shadow and not the new one. Can't remember if a compiler plugin would have caught an unused variable in that instance. 

Why do you want to avoid it?

Reply:
Well, I do not read code line by line, I scan it. I see a symbol declaration, I see its usage, I draw the obvious conclusion. Shadows may also be tricky when moving things around, like finding that the code still compiles even though you removed a variable it used.

But that's just my opinion.

Reply:
Good point about moving code around, I guess that's the other side of the coin I mentioned, which is accidentally using the original variable. 

Okay, renamed the variables and enabled compiler warnings for unused pattern matches and local values.